crates/foundry/evm/core/src/backend/cow.rs

```diff
@@ -265,6 +265,23 @@ impl<'a> DatabaseExt for CowBackend<'a> {
     fn has_cheatcode_access(&self, account: &Address) -> bool {
         self.backend.has_cheatcode_access(account)
     }
+
+    fn record_cheatcode_purity(&mut self, cheatcode_name: &'static str, is_pure: bool) {
+        // Only convert to mutable if we need to update.
+        if !is_pure
+            && !self
+                .backend
+                .inner
+                .impure_cheatcodes
+                .contains(cheatcode_name)
+        {
+            self.backend
+                .to_mut()
+                .inner
+                .impure_cheatcodes
+                .insert(cheatcode_name);
```

Comment:
Optional: I'm not sure about the performance implications of this branch, but you could reduce this to just an insert to avoid the branch prediction of the `contains` - if you're worried - especially as the `insert` likely has an `if` inside its implementation to determine whether a previous value existed.

Ideal would have been to use the `entry` API as it abstracts these two calls into one, but it's still nightly.

This suggestion is the same as your implementation in `crates/foundry/evm/core/src/backend/mod.rs`

Reply:
Thanks for the suggestion! The backend here is `Cow`, so the goal of the check is to avoid upgrading to an owned version if we don't actually have to insert.